app/ocertify.ml

```diff
@@ -101,25 +101,28 @@ let jump _ server_ip port (keyname, zone, dnskey) hostname csr key seed bits cer
        Error (`Msg (Fmt.strf "error %a while parsing TLSA reply" Dns_certify.pp_q_err e)))
   >>= fun send_update ->
   if send_update then
-    nsupdate_csr sock hostname keyname zone dnskey req >>= fun () ->
-    let rec request retries =
-      if retries = 0 then
-        Error (`Msg "failed to request certificate")
-      else
-        match query_certificate sock public_key hostname with
-        | Error `No_tlsa ->
-          Logs.warn (fun m -> m "still no tlsa, sleeping two more seconds");
-          Unix.sleep 2;
-          request (pred retries)
-        | Error (`Msg msg) ->
-          Logs.err (fun m -> m "error %s" msg);
-          Error (`Msg msg)
-        | Error ((`Decode _ | `Bad_reply _ | `Unexpected_reply _) as e) ->
-          Logs.err (fun m -> m "error %a while handling TLSA reply (retrying anyways)" Dns_certify.pp_q_err e);
-          request (pred retries)
-        | Ok x -> write_certificate x
-    in
-    request 10
+    match dns_key_opt with
+    | None -> Error (`Msg "no dnskey provided, but required for uploading CSR")
+    | Some (keyname, zone, dnskey) ->
+      nsupdate_csr sock hostname keyname zone dnskey req >>= fun () ->
+      let rec request retries =
+        if retries = 0 then
+          Error (`Msg "failed to request certificate")
+        else
+          match query_certificate sock public_key hostname with
+          | Error `No_tlsa ->
+            Logs.warn (fun m -> m "still no tlsa, sleeping two more seconds");
+            Unix.sleep 2;
```

Comment:
If I'm reading this correct the sleep will happen also in the case where `retries = 1`, then it will recurse with `retries = 0` which means it'll hit the `failed to request certificate` path, ie waiting for 2 seconds without anything productive to happen after that time, even if `query_certificate` now would return OK.

What do you think about something like this at the top of the pattern match instead?
```ocaml
| Error _ when retries = 0 -> Error (`Msg "failed to request certificate")
```

Reply:
i hope 5be954d satisfies you :)